external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala

```diff
@@ -0,0 +1,559 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.kinesis
+
+import scala.reflect.ClassTag
+
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.InitialPositionInStream
+import com.amazonaws.services.kinesis.model.Record
+
+import org.apache.spark.annotation.InterfaceStability
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.streaming.{Duration, StreamingContext}
+import org.apache.spark.streaming.api.java.JavaStreamingContext
+
+@InterfaceStability.Stable
+object KinesisDStream {
+  /**
+   * Builder for [[KinesisInputDStream]] instances.
+   *
+   * @param ssc StreamingContext to be used for creating DStreams
+   * @param streamName Name of Kinesis stream
+   * @param messageHandler Function for extracting DStream data type from KCL Record
+   *
+   * @since 2.2.0
+   */
+  @InterfaceStability.Stable
+  class Builder[T: ClassTag](
+      ssc: StreamingContext,
+      streamName: String,
+      messageHandler: Record => T) {
+    private var endpointUrl = DEFAULT_KINESIS_ENDPOINT_URL
+    private var regionName = DEFAULT_KINESIS_REGION_NAME
+    private var initialPositionInStream = DEFAULT_INITIAL_POSITION_IN_STREAM
+    private var checkpointAppName = ssc.sparkContext.appName
+    private var checkpointInterval = ssc.graph.batchDuration
+    private var storageLevel = DEFAULT_STORAGE_LEVEL
+    private var kinesisCredsProvider: SerializableCredentialsProvider = DefaultCredentialsProvider
+    private var dynamoDBCredsProvider: Option[SerializableCredentialsProvider] = None
+    private var cloudWatchCredsProvider: Option[SerializableCredentialsProvider] = None
+
+    /**
+     * Sets the AWS Kinesis endpoint URL. Defaults to "https://kinesis.us-east-1.amazonaws.com" if
+     * no custom value is specified
```

Comment:
I'll take a look and see. To be honest supplying both of these has always felt pretty redundant to me. The AWS SDK has changed a bit in how it handles endpoints and regions as well, so it may also be worth revisiting how ```KinesisReceiver``` uses these params.

Long term, it may also be nice to allow for different endpoints to be specified for Kinesis, DynamoDB and CloudWatch (I think the KCL should support this...)